src/rust/src/pkcs7.rs

```diff
@@ -162,6 +168,153 @@ fn encrypt_and_serialize<'p>(
+            let found_recipient_info = recipient_infos.find(|info| {
+                info.issuer_and_serial_number.serial_number == recipient_serial_number
+                    && info.issuer_and_serial_number.issuer == recipient_issuer
+            });
+
+            // Raise error when no recipient is found
+            // Unsure if this is the right exception to raise
+            let recipient_info = match found_recipient_info {
+                Some(info) => info,
+                None => {
+                    return Err(CryptographyError::from(
+                        exceptions::AttributeNotFound::new_err((
+                            "No recipient found that matches the given certificate.",
+                            exceptions::Reasons::UNSUPPORTED_X509,
+                        )),
+                    ));
+                }
+            };
+
+            // Decrypt the key using the private key
+            let padding = types::PKCS1V15.get(py)?.call0()?;
+            let key = private_key
+                .call_method1(
+                    pyo3::intern!(py, "decrypt"),
+                    (recipient_info.encrypted_key, &padding),
+                )?
+                .extract::<pyo3::pybacked::PyBackedBytes>()?;
+
+            // Get algorithm
+            // TODO: implement all the possible algorithms
+            let algorithm_identifier = enveloped_data
+                .encrypted_content_info
+                .content_encryption_algorithm;
+            let (algorithm, mode) = match algorithm_identifier.params {
+                AlgorithmParameters::Aes128Cbc(iv) => (
+                    types::AES128.get(py)?.call1((key,))?,
+                    types::CBC
+                        .get(py)?
+                        .call1((pyo3::types::PyBytes::new_bound(py, &iv),))?,
+                ),
+                _ => {
+                    return Err(CryptographyError::from(
+                        exceptions::UnsupportedAlgorithm::new_err((
+                            "Only AES-128-CBC is currently supported for decryption.",
+                            exceptions::Reasons::UNSUPPORTED_SERIALIZATION,
+                        )),
+                    ));
+                }
+            };
+
+            // Decrypt the content using the key and proper algorithm
+            let encrypted_content = enveloped_data
+                .encrypted_content_info
+                .encrypted_content
+                .unwrap();
+            let decrypted_content = symmetric_decrypt(py, algorithm, mode, encrypted_content)?;
+            pyo3::types::PyBytes::new_bound(py, decrypted_content.as_slice())
+        }
+        _ => {
+            return Err(CryptographyError::from(
+                exceptions::UnsupportedAlgorithm::new_err((
+                    "Only EnvelopedData structures are currently supported.",
+                    exceptions::Reasons::UNSUPPORTED_SERIALIZATION,
+                )),
```

Comment:
Is this unsupported, or just a ValueError, because the data is in the wrong format?

Reply:
As far as I understand from S/MIME 3.2 [RFC](https://datatracker.ietf.org/doc/html/rfc5751), it seems that only enveloped data can be decrypted. So I'm passing this as a ValueError then. I guess I initially thought other formats could be decrypted.